Only set ip6tables filter-FORWARD DROP if necessary

robmry opened the pull request:
**- What I did**

- Fix https://github.com/moby/moby/issues/48365

The current behaviour for IPv4, in [setupIPForwarding](https://github.com/moby/moby/blob/9fca5e7b6ff9aa29f85930d616565d5e1e09fad3/libnetwork/drivers/bridge/setup_ip_forwarding.go#L27-L74), when the daemon runs with `"iptables": true` is:

  - if the daemon is running with `--ip-forward` (the default), and
  - if the kernel's `net.ipv4.ip_forward = 0` when the daemon starts, then
  - enable `net.ipv4.ip_forward` and set the `FORWARD` chain's default policy to `DROP`.

The current behaviour for IPv6, when `"ip6tables": true` is to unconditionally set the filter `FORWARD` chain's policy to `DROP`. Because `ip6tables` has been enabled by-default in 27.x releases, this has caused issues on hosts that relied on an `ACCEPT` policy to make other things work. (For example, if `ip6tables` wasn't explicitly disabled, even if no IPv6 networks were created, dockerd set the ip6tables filter-FORWARD policy to `DROP`).

**- How I did it**

Still set the filter-FORWARD policy to DROP if enabling IP forwarding.
- This happens on startup for IPv4, when the default bridge network is configured. For IPv6, it happens when an IPv6 network is created.
- If IP forwarding was enabled by something else before docker starts, the policy is left as it was. This is the old behaviour for IPv4, but a change for IPv6.
- Added daemon option `--ip-filter-forward-drop=<bool>` ... the default is `true` and means that the policy will be set to drop if the daemon enables IP forwarding (as described above). If `false` docker will not update the policy:
  - ```
      --ip-forward                              Enable IP forwarding in system configuration (default true)
      --ip-forward-no-drop                      Do not set the filter-FORWARD policy to DROP when enabling IP forwarding
    ```

**- How to verify it**

New integration tests.

**- Description for the changelog**
```markdown changelog
- docker will now only set the `ip6tables` policy for the `FORWARD` chain in the `filter` table to `DROP` if
  it enables IP forwarding on the host itself (sysctls `net.ipv6.conf.all.forwarding` and
  `net.ipv6.conf.default.forwarding`). This is now aligned with existing IPv4 behaviour.
  - If IPv6 forwarding is enabled on your host, but you were depending on docker to set the ip6tables
    filter-FORWARD policy to `DROP`, you may need to update your host's configuration to make sure it
    is secure.
  - If docker enables IP forwarding on your host and you do not want it to set the filter-FORWARD policy
    to `DROP`, use daemon command line option `--ip-forward-no-drop` or `"ip-forward-no-drop": true`
    in `daemon.json`.
- Direct routed access to container ports that are not exposed using `-p`/`--publish` is now blocked
   in the `DOCKER` chain.
  - If the default filter-FORWARD policy was previously left at `ACCEPT` on your host, and direct routed
    access to a container's unpublished ports from a remote host is still required, options are:
    - Publish the ports you need.
    - Re-create the network with `--gateway_mode_ipv4=routed` and/or `--gateway_mode_ipv6=routed`.
      - No NAT rules will be set up to map host addresses/ports to containers, but published ports will be
        accessible from remote hosts if routing has been set up in the network.
    - (FIXME - mention `nat-unprotected`, to be implemented in an upcoming PR.)
- Container ports published to host addresses will continue to be accessible via those host
  addresses, using NAT or the userland proxy.
- Unpublished container ports continue to be directly accessible from the docker host via the
  container's IP address.
```

robmry commented:
Hmm, apologies - some last minute second thoughts about this ...

If we do enable forwarding for IPv4, and don't set the policy to DROP - this PR means the bridge networks will be protected ... but enabling forwarding might expose other-stuff on the host. We'd have to let users know they need to set the policy to DROP, if needed. But, a lot won't know whether it's needed, or won't notice the warning.

Perhaps we should continue to set the policy to DROP if we enable IPv4 forwarding - and do the same for IPv6; only set the policy if we enable forwarding, rather than unconditionally if ip6tables=true.

Then we'd need to document that "if you don't want us to break other things by setting the DROP policy, you need to enable ip forwarding before starting the daemon"?

We could also add daemon options to say what the filter-FORWARD policies should be for v4/v6 (or, alternatively and maybe-better, whether to leave them alone). Then, there'd be no need for users to work out how to manually configure IP forwarding before starting the daemon.

This PR will still be a big improvement because an ACCEPT policy will no longer leave bridge networks exposed, but more secure by-default - even if the default breaks things. We can write some loud documentation about what sort of things might break with a DROP policy, and how to stop docker from setting it.

robmry committed:
Add daemon option --ip-forward-no-drop

The daemon no longer depends on the iptables/ip6tables filter-FORWARD
chain's policy being DROP in order to implement its port filtering
rules.

However, if the daemon enables IP forwarding in the host's system
config, by default it will set the policy to DROP to avoid potential
security issues for other applications/networks.

If docker does need to enable IP forwarding, but other applications
on the host require filter-FORWARD's policies to be ACCEPT, this
option can be used to tell the daemon to leave the policy unchanged.
(Equivalent to enabling IP forwarding before starting the daemon,
but without needing to do that.)

Signed-off-by: Rob Murray <rob.murray@docker.com>

robmry committed:
Add TestFilterForwardPolicy

Signed-off-by: Rob Murray <rob.murray@docker.com>

robmry committed:
Move integration/network/bridge_test.go to bridge subdir

Also make it Linux-only, as the bridge driver is Linux only and
all of the tests had skips for Windows.

Signed-off-by: Rob Murray <rob.murray@docker.com>

robmry commented:
> Perhaps we should continue to set the policy to DROP if we enable IPv4 forwarding - and do the same for IPv6; only set the policy if we enable forwarding, rather than unconditionally if ip6tables=true.

That's done - the handling for IPv6 is now the same as it was for IPv4 ... the filter-FORWARD policy is only set to DROP if the daemon enables IP forwarding for that address family.

Also added daemon option `--ip-filter-forward-drop`, if set to `false` the daemon will not set the policy to DROP (for either address family) even if it enabled IP forwarding.

This part of the change alone is probably enough to fix https://github.com/moby/moby/issues/48365 in most cases ... enabling `ip6tables` by default caused us to set its policy to DROP, even if no IPv6-enabled bridge network was ever created and IPv6 forwarding remained disabled. It won't do that any more.

Changes to remove the dependency on the filter-FORWARD policy to implement port filtering were part of this PR, now they're in https://github.com/moby/moby/pull/48724.

akerouanton commented on libnetwork/drivers/bridge/setup_ip_forwarding.go:
```diff
@@ -33,42 +27,93 @@ func setupIPForwarding(enableIPTables bool, enableIP6Tables bool) error {
+	}()
+
+	// Get current IPv6 all forwarding setup
+	ipv6ForwardDataAll, err := os.ReadFile(ipv6ForwardConfAll)
```
Maybe it'd be benefitial here too to read both files, early-return if both are already set, and if not then start with ipt default policy and then set both unconditionnally. That'd make the code simpler I think.

robmry replied:
As above - we'd still need to undo anything done so-far, so I'm not sure it'd help?

But I've put the undo-sysctl code in the right place now (oops), so perhaps it makes more sense?

thaJeztah replied:
It's existing code, but curious; we _fail_ this function if we fail to read, but we _don't_ return an error when failing to update it? Could that have situations where we would not actually be updating the config? (Are those save to be ignored?).

thaJeztah replied:
Looks like for IPv4 we have a separate utility (`configureIPForwarding()`). Wondering if (not for this PR), we should have that utility handle the "changed" or "not changed" case, e.g. something like;

```go
func configureIPForwarding(file string, enable bool) (changed bool, _ error) {
	data, err := os.ReadFile(file)
	if err != nil {
		return false, fmt.Errorf("cannot read IP forwarding setup: %w", err)
	}
	if enabled := data[0] == '1'; enabled == enable {
		return false, nil
	}
	var val byte
	if enable {
		val = '1'
	}
	if err := os.WriteFile(file, []byte{val, '\n'}, 0o644); err != nil {
		return false, fmt.Errorf("cannot write IP forwarding setup: %w", err)
	}
	return true, nil
}
```

That would return a boolean to indicate we changed anything, but an error when failing to either read or update (but that depends a bit on my previous comment).

The above wraps the errors, so we could detect `os.IsnotFound()` situations (if those are ok to be ignored).

robmry replied:
Right, yes ... things won't work properly if IPv6 forwarding is needed and it can't be enabled, and I can't see why the error handling should be different to IPv4 - so I'll do that, thank you.

akerouanton commented on libnetwork/drivers/bridge/setup_ip_forwarding_test.go:
```diff
@@ -6,20 +6,23 @@ import (
 
 func TestSetupIPForwarding(t *testing.T) {
-	// Read current setting and ensure the original value gets restored
+	defer netnsutils.SetupTestOSContext(t)()
```
:+1:

akerouanton commented on integration/network/bridge/bridge_linux_test.go:
```diff
@@ -193,3 +196,124 @@ func TestIPRangeAt64BitLimit(t *testing.T) {
+// host, it also sets the iptables filter-FORWARD policy to DROP (unless it's
+// told not to).
+func TestFilterForwardPolicy(t *testing.T) {
+	skip.If(t, testEnv.DaemonInfo.OSType == "windows", "no bridge driver or iptables on Windows")
```
```suggestion
```

That's not needed, we're in `bridge_linux_test.go`.

robmry replied:
Good point!

robmry commented on libnetwork/drivers/bridge/setup_ip_forwarding.go:
```diff
@@ -33,42 +27,93 @@ func setupIPForwarding(enableIPTables bool, enableIP6Tables bool) error {
+
+		// When enabling ip_forward set the default policy on forward chain to drop.
+		if wantFilterForwardDrop {
+			if err := setFilterForwardDrop(iptables.IPv4); err != nil {
```
Setting the sysctl will fail if, for example, "/proc/sys/net" is mounted read-only and, by the time this is happening, if iptables is enabled we'll already have tried to mess with the rules (so the sysctl setting is perhaps more likely to fail).

Swapping them would just mean having to undo the policy change if the sysctl update fails. So I'm not sure it'd help. (But perhaps I missed the point?!)

robmry committed:
Move setupIPv6Forwarding() to setup_ip_forwarding.go

Signed-off-by: Rob Murray <rob.murray@docker.com>

thaJeztah commented on libnetwork/drivers/bridge/setup_ip_forwarding.go:
```diff
@@ -12,78 +12,65 @@ import (
+		defer func() {
+			if retErr != nil {
+				if err := os.WriteFile(ipv6ForwardConfDefault, []byte{'0', '\n'}, forwardConfPerm); err != nil {
+					log.G(context.TODO()).WithError(err).Error("Disabling IPv6 default forwarding failed")
```
The `WriteFile` error above is ignored (only logged), but here we try to write to it again in the `defer`. Wondering if the first write should be returned as well (there's nothing to "undo" if we failed to write, and it's also likely that if we failed to write the first time that it'll fail the second time?)

The `configUpdated` also may be incorrect here, because if we failed to write, then no config was updated 🤔

robmry replied:
Right, yes - I've aligned IPv4/IPv6 error handling as you suggested above.

thaJeztah commented on libnetwork/drivers/bridge/setup_ip_forwarding_test.go:
```diff
@@ -39,15 +42,8 @@ func readCurrentIPForwardingSetting(t *testing.T) []byte {
 }
 
 func writeIPForwardingSetting(t *testing.T, chars []byte) {
```
This utility is only used in a single test; possibly (with the other idea of having a utility for this), we could reuse that utility in the tests as well;

```go
_, err := configureIPForwarding(ipv4ForwardConf, true)
if err != nil {
    t.Fatalf(err)
}
```

robmry replied:
Done.

thaJeztah commented on libnetwork/drivers/bridge/setup_ipv6_linux.go:
```diff
@@ -6,15 +6,11 @@ import (
 	"os"
 )
 
+const disableIPv6ConfPerm = 0o644
```
Perhaps use the same const as for IPv4 (`forwardConfPerm`) or even consider removing the const entirely.

robmry replied:
Removed.

thaJeztah commented on daemon/daemon_unix.go:
```diff
@@ -915,12 +915,13 @@ func setHostGatewayIP(controller *libnetwork.Controller, config *config.Config)
-			"UserlandProxyPath":   config.BridgeConfig.UserlandProxyPath,
-			"Rootless":            config.Rootless,
+			"EnableIPForwarding":      config.BridgeConfig.EnableIPForward,
+			"EnableFilterForwardDrop": config.BridgeConfig.EnableFilterForwardDrop,
```
I don't think we have docs for this in [man/dockerd.8.md](https://github.com/moby/moby/blob/6ac445c42badc98b416baa32cca5d6f8b712b359/man/dockerd.8.md) wondering if we should look at that in a follow-up 🤔 

Is this an option that can be "live reloaded" ? (i.e., enabled/disabled when reloading the daemon config?)

robmry replied:
Added to the manpage.

For config reload ... it's not, like `--ip-forward` and `--iptables`/`--ip6tables`. I think they're all similar in that they affect host config when a network's created, and that host config won't be un-done if the daemon config is changed.

Perhaps someone might want to fix the filter-FORWARD policy manually after creating a network, then reload daemon config so that it's not changed again when the next network is created. But, I think the same applies to IP forwarding and iptables and it's probably best that they all have the same behaviour?

thaJeztah commented on cmd/dockerd/config_unix.go:
```diff
@@ -27,7 +27,8 @@ func installConfigFlags(conf *config.Config, flags *pflag.FlagSet) {
 	flags.BoolVar(&conf.BridgeConfig.EnableIP6Tables, "ip6tables", true, "Enable addition of ip6tables rules")
-	flags.BoolVar(&conf.BridgeConfig.EnableIPForward, "ip-forward", true, "Enable net.ipv4.ip_forward")
+	flags.BoolVar(&conf.BridgeConfig.EnableIPForward, "ip-forward", true, "Enable IP forwarding in system configuration")
+	flags.BoolVar(&conf.BridgeConfig.EnableFilterForwardDrop, "ip-filter-forward-drop", true, "Set the filter-FORWARD policy to DROP when enabling IP forwarding")
```
This needs an update in [man/dockerd.8.md](https://github.com/moby/moby/blob/6ac445c42badc98b416baa32cca5d6f8b712b359/man/dockerd.8.md) and the corresponding daemon reference docs in [docker/cli](https://github.com/docker/cli/blob/2995631498fc00463e387bf60e45947db5c6e025/docs/reference/dockerd.md)

Also curious; should enabling this without the other option be an error-condition? Or is it fine to ignore?

robmry replied:
Added to the manpage.

I think it's ok to ignore - it says it'll only do-something if the daemon enables forwarding itself, and it won't if `--ip-forward=false` (or if iptables/ip6tables is disabled).

thaJeztah replied:
Oh, I didn't look at this closely; I just realised we're introducing a new boolean flag, but with `true` as default. I realise we already have some those, but if we consider this default to be `true` in the foreseeable future, then perhaps we should have `false` as default and name the flag accordingly.

Having `false` makes the UX more usable (instead of `--boolean-flag=false`), but also can help prevent issues when reading config from `daemon.json` where a missing field would default to "false" when unmarshaling.


(just a quick comment; about to jump in a call)

robmry replied:
Right, yes - the closely related `--iptables`, `--ip6tables` and `--ip-forward` flags all default to `true` ... I guessed that was probably deliberate! They're all "batteries included" flags ... you need to leave them enabled to make things work, unless you know about the innards and have a good reason to turn them off.

So, I'm not sure a default-false option fits well with those others and, if we do change it, it'd need to be called something like `--no-ip-filter-forward-drop`. Then it's a bit too double-negative-y, "`--no-ip-filter-forward-drop=false`" means it will set the filter-FORWARD policy to DROP if it needs to.

(Unless anyone has a better idea for the flag name, it is already a bit cumbersome ... but I thought it was good to start it with "`--ip...`" so that it's close to those other flags in help text, and I think it'll still make sense for native nftables.)

I don't think there's an issue with merging CLI and daemon.json options, in the same way as for the other options - no option is allowed to be specified in both, if it's in the JSON it'll override the CLI's default. If not present in the JSON or the CLI, the CLI's default wins.

robmry replied:
After further discussion in Slack, I've changed the option to `--ip-forward-no-drop`.

No engineers were harmed in the making of this decision!

robmry committed:
Modify IP forwarding and filter-FORWARD DROP setup.

Before this change, for IPv4:
- sysctl net.ipv4.ip_forward was enabled during bridge driver
  initialisation, if:
  - not already enabled
  - ip-forward=true, and
  - iptables=true.
- the filter-FORWARD chain's policy was set to DROP, if the daemon
  updated the sysctl.
- if setting the policy failed, the sysctl change was reverted.

But, for IPv6:
- sysctls net.ipv6.conf.[default|all].forwarding were both enabled
  when creating the first IPv6-enabled network, if:
  - they weren't already enabled,
  - ip-forward=true, and
  - ip6tables=true.
- the filter-FORWARD chain's policy was set to DROP when creating
  an IPv4 enabled bridge network (inc. the default bridge), if:
  - ip6tables=true.
  - (this happened whether or not the daemon would ever enable
    IPv6 forwarding, or even create an IPv6 network.)

The bridge driver no longer needs the default policy to be DROP to
implement its own port-filtering rules. But, enabling IP forwarding
without setting the filter-FORWARD policy to DROP would potentially
be a security risk.

This change aligns IPv4 and IPv6 behaviours:
- only try to set the sysctls when creating a bridge network that
  needs them (for IPv4, that's still during daemon init because
  the default bridge is IPv4 enabled).
- only check/set the filter-FORWARD policy after updating sysctls.
- if the filter-FORWARD policy can't be set, treat it as an error
  and revert sysctl changes.

We enabled ip6tables by default in 27.0. Setting the filter-FORWARD
policy to DROP even when no IPv6 enabled network was created
caused issues for some users. In particular, those running with
iptables=false suddenly got the IPv6 DROP policy enabled (which
broke unrelated services on the host). This change solves that by
only setting the policy when necessary.

Signed-off-by: Rob Murray <rob.murray@docker.com>

robmry commented:
Rebased - and added a check of `config.EnableIPv4` before enabling IPv4 forwarding, now that flag has been merged to master.

thaJeztah commented on man/dockerd.8.md:
```diff
@@ -44,6 +44,7 @@ dockerd - Enable daemon mode
 [**--insecure-registry**[=*[]*]]
 [**--ip**[=*0.0.0.0*]]
 [**--ip-forward**[=**true**]]
+[**--ip-filter-forward-drop**[=**true**]]
```
Looks like you forgot updating this one 🙈

thaJeztah commented on man/dockerd.8.md:
```diff
@@ -289,11 +290,20 @@ unix://[/path/to/socket] to use.
+**--ip-forward-no-drop**=**true**|**false**
+  When **false**, the default, if Docker enables IP forwarding itself (see
+  **--ip-forward**), and **--iptables** or **--ip6tables** are enabled, it
+  will also set the default policy for the FORWARD chain in the iptables
```
Nit; avoid future tense; this could be "it also sets the ..."

robmry replied:
Done - there are a few existing examples of that in the text, I guess we should tidy those too (`--default-address-pool`, `--ip-forward`, `--ipv6`, `--userns-remap`, `zfs.fsname`).

thaJeztah replied:
Ah, yes, we could make a pass at doing so. We also still need to change our packaging to use the one from this repository in stead of the copy that still lives in docker/cli 🙈

robmry replied:
Oh, right, there are two `docker.md.8`'s! So we should be keeping them in step for now?

thaJeztah commented on man/dockerd.8.md:
```diff
@@ -289,11 +290,20 @@ unix://[/path/to/socket] to use.
+  will also set the default policy for the FORWARD chain in the iptables
+  or ip6tables filter table to DROP.
+
+  When **true**, and when IP forwarding is already enabled, Docker will
+  not modify the default policy of the FORWARD chain.
```
Same here; "Docker does not modify ..."

thaJeztah reviewed (approved):
LGTM, but found some minor issues in the fourth commit; so ok to merge after those are addressed
